lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -282,42 +592,139 @@ public void OnCompleted()
+        }
+
+        /// <summary>
+        /// If all the tasks are in final state, if the system is recoverable, start recovery
+        /// else, change the system state to Fail then take Fail action
+        /// </summary>
+        private void TryRecovery()
+        {
+            if (_taskManager.AllInFinalState())
+            {
+                if (IsRecoverable())
+                {
+                    _isFirstTry = false;
+                    StartAction();
+                }
+                else
+                {
+                    Logger.Log(Level.Warning, "The system is not recoverable, change the state to Fail.");
+                    _systemState.MoveNext(SystemStateEvent.NotRecoverable);
+                    FailAction();
+                }
+            }
         }
 
-        private string GetTaskIdByEvaluatorId(string evaluatorId)
+        private string GetMapperTaskIdByEvaluatorId(string evaluatorId)
         {
             return string.Format("{0}-{1}-Version0",
                 IMRUConstants.MapTaskPrefix,
                 _serviceAndContextConfigurationProvider.GetPartitionIdByEvaluatorId(evaluatorId));
         }
 
+        private void DoneAction()
+        {
+            ShutDownAllEvaluators();
+        }
+
+        private void FailAction()
+        {
+            ShutDownAllEvaluators();
+        }
+
         /// <summary>
-        /// Shuts down evaluators once all completed task messages are received
+        /// Shuts down evaluators
         /// </summary>
         private void ShutDownAllEvaluators()
         {
-            foreach (var task in _completedTasks)
+            foreach (var context in _contextManager.ActiveContexts)
             {
-                Logger.Log(Level.Info, string.Format("Disposing task: {0}", task.Id));
-                task.ActiveContext.Dispose();
+                Logger.Log(Level.Verbose, "Disposing active context: {0}", context.Id);
+                context.Dispose();
             }
         }
 
         /// <summary>
+        /// Resets Failed Evaluators
+        /// Changes state to WaitingForEvaluator if it is in recovery.
+        /// If it is first time, request both master and map evaluators
+        /// If it is in recovery, based on the count of failed evaluators, requests missing map evaluators
+        /// </summary>
+        private void StartAction()
+        {
+            lock (_lock)
+            {
+                if (!_isFirstTry)
+                {
+                    _numberOfRetryForFaultTolerant++;
+                    Logger.Log(Level.Info, "Start recovery with _numberOfRetryForFaultTolerant:" + _numberOfRetryForFaultTolerant);
+                    _systemState.MoveNext(SystemStateEvent.Recover);
+                }
+
+                var mappersToRequest = _isFirstTry ? _totalMappers : _evaluatorManager.NumberofFailedMappers();
+                _evaluatorManager.ResetFailedEvaluators();
+
+                if (!_isFirstTry && mappersToRequest == 0)
```

Comment:
If a task fails for some reason, not Evaluator failure, we will  close rest of the tasks and then reach here.